File: split_files.js
const fs = require('fs');
const path = require('path');

const rootDir = __dirname;
const indexPath = path.join(rootDir, 'index.html');
const cssDir = path.join(rootDir, 'css');
const jsDir = path.join(rootDir, 'js');

// Create directories if they don't exist
if (!fs.existsSync(cssDir)) fs.mkdirSync(cssDir, { recursive: true });
if (!fs.existsSync(jsDir)) fs.mkdirSync(jsDir, { recursive: true });

// Read index.html content
let content = fs.readFileSync(indexPath, 'utf-8');

// 1. Extract Preloader CSS
const preTag = '<style id="crt-preloader-styles">';
const preStart = content.indexOf(preTag);
if (preStart === -1) {
    console.error("Could not find preloader styles start tag");
    process.exit(1);
}
const preEnd = content.indexOf('</style>', preStart);
const preCss = content.substring(preStart + preTag.length, preEnd).trim();

// 2. Extract Main CSS
const mainCssTag = '<style>';
const mainCssStart = content.indexOf(mainCssTag, preEnd);
if (mainCssStart === -1) {
    console.error("Could not find main styles start tag");
    process.exit(1);
}
const mainCssEnd = content.indexOf('</style>', mainCssStart);
const mainCss = content.substring(mainCssStart + mainCssTag.length, mainCssEnd).trim();

// 3. Extract Main JS
const mainJsTag = '<script>';
const mainJsStart = content.indexOf(mainJsTag, mainCssEnd);
if (mainJsStart === -1) {
    console.error("Could not find main script start tag");
    process.exit(1);
}
const mainJsEnd = content.indexOf('</script>', mainJsStart);
const mainJs = content.substring(mainJsStart + mainJsTag.length, mainJsEnd).trim();

// Write sub-files
fs.writeFileSync(path.join(cssDir, 'preloader.css'), preCss, 'utf-8');
fs.writeFileSync(path.join(cssDir, 'main.css'), mainCss, 'utf-8');
fs.writeFileSync(path.join(jsDir, 'script.js'), mainJs, 'utf-8');

// Construct new HTML content
const newHtml = 
    content.substring(0, preStart) +
    '<link rel="stylesheet" href="css/preloader.css" id="crt-preloader-styles">' +
    content.substring(preEnd + '</style>'.length, mainCssStart) +
    '<link rel="stylesheet" href="css/main.css">' +
    content.substring(mainCssEnd + '</style>'.length, mainJsStart) +
    '<script defer src="js/script.js"></script>' +
    content.substring(mainJsEnd + '</script>'.length);

fs.writeFileSync(indexPath, newHtml, 'utf-8');

console.log("Successfully split monolithic index.html into modular sub-files via Node.js!");

File: split_files.py
import os

# Define file paths
root_dir = os.path.dirname(os.path.abspath(__file__))
index_path = os.path.join(root_dir, 'index.html')
css_dir = os.path.join(root_dir, 'css')
js_dir = os.path.join(root_dir, 'js')

# Create directories
os.makedirs(css_dir, exist_ok=True)
os.makedirs(js_dir, exist_ok=True)

# Read index.html content
with open(index_path, 'r', encoding='utf-8') as f:
    content = f.read()

# 1. Extract Preloader CSS
pre_tag = '<style id="crt-preloader-styles">'
pre_start = content.find(pre_tag)
if pre_start == -1:
    raise ValueError("Could not find preloader styles start tag")
pre_end = content.find('</style>', pre_start)
pre_css = content[pre_start + len(pre_tag):pre_end].strip()

# 2. Extract Main CSS
main_css_tag = '<style>'
main_css_start = content.find(main_css_tag, pre_end)
if main_css_start == -1:
    raise ValueError("Could not find main styles start tag")
main_css_end = content.find('</style>', main_css_start)
main_css = content[main_css_start + len(main_css_tag):main_css_end].strip()

# 3. Extract Main JS
main_js_tag = '<script>'
main_js_start = content.find(main_js_tag, main_css_end)
if main_js_start == -1:
    raise ValueError("Could not find main script start tag")
main_js_end = content.find('</script>', main_js_start)
main_js = content[main_js_start + len(main_js_tag):main_js_end].strip()

# Write sub-files
with open(os.path.join(css_dir, 'preloader.css'), 'w', encoding='utf-8') as f:
    f.write(pre_css)

with open(os.path.join(css_dir, 'main.css'), 'w', encoding='utf-8') as f:
    f.write(main_css)

with open(os.path.join(js_dir, 'script.js'), 'w', encoding='utf-8') as f:
    f.write(main_js)

# Construct new HTML content
new_html = (
    content[:pre_start] +
    '<link rel="stylesheet" href="css/preloader.css" id="crt-preloader-styles">' +
    content[pre_end + len('</style>'):main_css_start] +
    '<link rel="stylesheet" href="css/main.css">' +
    content[main_css_end + len('</style>'):main_js_start] +
    '<script defer src="js/script.js"></script>' +
    content[main_js_end + len('</script>'):]
)

with open(index_path, 'w', encoding='utf-8') as f:
    f.write(new_html)

print("Successfully split monolithic index.html into modular sub-files dynamically!")
